### ISSUE TITLE
Some frame dimensions cause desynchronizations

### ISSUE THREAD

**Issue by rzumer** [xiph/rav1e#385]
The `dimensions` encode-decode test reveals that very small (4x4 and below) and very high (2048x2048 and above) frame sizes induce desynchronization between the values signaled to the encoder and the resulting decoded values. The smallest frame size verified to be coded correctly is 8x8, and the largest 1024x1024, so the actual lower and upper limits have yet to be verified.

In addition, some frame dimensions that are not divisible by 2<sup>4</sup> (16) either desynchronize on edges (e.g. 264x128, where the width is divisible by 2<sup>3</sup>) or cause invalid memory reference errors (e.g. 260x128, where the width is divisible by 2<sup>2</sup>).

These observations are based on decoding test results.

**Comment by tdaede** [comment]
The AV1 spec officially only supports 8x8 video, however I think there's nothing stopping you from encoding a smaller one. The max tile size is 4kx2k, but 2048x2048 should just squeeze in that...

**Comment by rzumer** [comment]
The decoder might just not be concerned with decoding lower frame sizes correctly, in that case. Samples are decoded correctly until the halfway point within a frame.

With high frame sizes, the decoded buffer is filled with zeroes instead of the expected sample values.

**Comment by ycho** [comment]
I've always tested with subset1-y4m/08-2011._Panthera_tigris_tigris_-_Texas_Park_-_Lanzarote_-TP04.y4m, which has size 1276x850 = wxh, and it is x2 but not x4.
But rav1e did not cause desync.

**Comment by ycho** [comment]
With my patch for adding new test, https://github.com/ycho/rav1e/commit/4bc95cf39efd038f9bfe968579151be2bdc1f66b,
--------------------------------
If wxh = 64x64, and run "cargo test --features=decode_test -- --ignored"
I've got:

running 4 tests
test test_encode_decode::dimensions ... FAILED
test test_encode_decode::quantizer ... FAILED
test test_encode_decode::speed ... FAILED
test test_encode_decode::odd_size_frame_with_full_rdo ... ok

failures:

---- test_encode_decode::dimensions stdout ----
	Encoding 8x8 speed 4 quantizer 100
Encoding frame 0
thread 'test_encode_decode::dimensions' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- test_encode_decode::quantizer stdout ----
	Encoding 64x80 speed 4 quantizer 80
Encoding frame 0
thread 'test_encode_decode::quantizer' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20

---- test_encode_decode::speed stdout ----
	Encoding 64x80 speed 0 quantizer 100
Encoding frame 0
thread 'test_encode_decode::speed' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20


failures:
    test_encode_decode::dimensions
    test_encode_decode::quantizer
    test_encode_decode::speed

test result: FAILED. 1 passed; 3 failed; 0 ignored; 0 measured; 34 filtered out

error: test failed, to rerun pass '--lib'
----------------------------------
If I set wxh = 63x63,

   Compiling rav1e v0.1.0 (file:///Users/yushin/workspace/rav1e)
    Finished dev [unoptimized + debuginfo] target(s) in 4.08s
     Running target/debug/deps/rav1e-4915d6be767af556

running 4 tests
test test_encode_decode::dimensions ... FAILED
test test_encode_decode::quantizer ... FAILED
test test_encode_decode::odd_size_frame_with_full_rdo ... FAILED
test test_encode_decode::speed ... FAILED

failures:

---- test_encode_decode::dimensions stdout ----
	Encoding 8x8 speed 4 quantizer 100
Encoding frame 0
thread 'test_encode_decode::dimensions' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- test_encode_decode::quantizer stdout ----
	Encoding 64x80 speed 4 quantizer 80
Encoding frame 0
thread 'test_encode_decode::quantizer' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20

---- test_encode_decode::odd_size_frame_with_full_rdo stdout ----
	Encoding 63x63 speed 0 quantizer 100
Encoding frame 0
Encoded.
Decoding frame 0
Decoded. -> 0
Retrieving frame
Retrieved.
thread 'test_encode_decode::odd_size_frame_with_full_rdo' panicked at 'assertion failed: `(left == right)`
  left: `[128, 160, 72, 84, 192, 166, 169, 139, 220, 74, 116, 188, 108, 17, 191, 183, 155, 81, 157, 117, 205, 75, 21, 50, 39, 157, 221, 115, 59, 85, 171, 74, 44, 107, 177, 29, 142, 116, 172, 164, 6, 29, 41, 230, 121, 192, 96, 65, 24, 213, 204, 71, 199, 251, 193, 170, 57, 239, 233, 17, 35, 106, 231, 54]`,
 right: `[128, 160, 72, 84, 192, 166, 169, 139, 220, 74, 116, 188, 108, 17, 191, 183, 155, 81, 157, 117, 205, 75, 21, 50, 39, 157, 221, 115, 59, 85, 171, 74, 44, 107, 177, 29, 142, 116, 172, 164, 6, 29, 41, 230, 121, 192, 96, 65, 24, 213, 204, 71, 199, 251, 193, 170, 57, 239, 233, 17, 35, 106, 231, 231]`', src/lib.rs:2114:13

---- test_encode_decode::speed stdout ----
	Encoding 64x80 speed 0 quantizer 100
Encoding frame 0
thread 'test_encode_decode::speed' panicked at 'attempt to multiply with overflow', src/cdef.rs:58:20


failures:
    test_encode_decode::dimensions
    test_encode_decode::odd_size_frame_with_full_rdo
    test_encode_decode::quantizer
    test_encode_decode::speed

test result: FAILED. 0 passed; 4 failed; 0 ignored; 0 measured; 34 filtered out

error: test failed, to rerun pass '--lib'
